### PR TITLE
feat: Functionality to cache and reuse dependencies before build

### LINKED ISSUE #1118: [FEATURE] Enable caching of dependencies to save build minutes
### Description

- In `.github/workflows/build.yaml` we can cache the dependencies based on changes in `package-lock.json` .
- This will help make our build much faster.
- Since for free tier, only 2000 build minutes/month are allotted on github actions servers, if in future the project becomes very large, we can take advantage of caching and keep the build times low.

#### I would like to work on this issue
- I will create hashes for the package-lock and setup lookups based on hashes to help fetch the cache if the dependencies are same as before.

**Please assign me this issue if found valid** 🙂 

### Screenshots

_No response_

### Additional information

- Also I found that there was a lot of code duplication among  `.github/workflows/build.yaml` file and the other workflows as well such as checking out to the repo in the VM and then installing all dependencies. 
- I do know a couple ways that might help but not sure about them. I can raise new issue for this and research on it as well or we can discuss with other community members.
- A couple ways that I have for this are making a single reusable template(composite actions) or combining jobs just have separate steps but that would compromise parallelism but then we don't have to setup a new environment again.
- Let me know if this sounds valid, I'll raise a new issue for it then and research on it a bit more.

### PR DESCRIPTION
**Pull request by Pranav016**
## Fixes Issue
Closes #1118 

## Changes proposed
- Add functionality to cache dependencies and reuse the cache if the dependencies haven't change.
- This makes the build process much faster (we can build within seconds)
- Since for free tier, only 2000 build minutes/month are allotted on GitHub action servers, if in future the project becomes very large, we can take advantage of caching and keep the build times low.

## Check List (Check all the applicable boxes) 

- [x] My code follows the code style of this project.
- [ ] My change requires changes to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [ ] This PR does not contain plagiarized content.
- [x] The title of my pull request is a short description of the requested changes.

<a href="https://gitpod.io/#https://github.com/EddieHubCommunity/LinkFree/pull/1124"></a>

### CODE REVIEW

**Comment by Pranav016** [comment]
@eddiejaoude can you pls review this PR? 🙂

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -23,6 +30,13 @@ jobs:
       - uses: actions/setup-node@v2
         with:
           node-version: '16'
+      - name: cache lookup
+        uses: actions/cache@v2
```
Comment (nhcarrigan) [inline .github/workflows/build.yml:34]:
```suggestion
      - name: cache lookup
        id: cache
        uses: actions/cache@v2
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -9,6 +9,13 @@ jobs:
...
+        with:
+          path: ~/.npm
+          key: ${{ runner.os }}-node-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            ${{ runner.os }}-node-
       - name: install dependencies
         run: npm ci
```
Comment (nhcarrigan) [inline .github/workflows/build.yml:20]:
```suggestion
      - name: install dependencies
        if: steps.cache.outputs.cache-hit != 'true'
        run: npm ci
```

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -23,6 +30,13 @@ jobs:
...
```
Comment (nhcarrigan) [inline .github/workflows/build.yml:41]:
```suggestion
      - name: install dependencies
        if: steps.cache.outputs.cache-hit != 'true'
        run: npm ci
```

**Review by nhcarrigan — CHANGES_REQUESTED** [review]
My understanding is we actually need to check if the cache step has a hit, and skip the install step if it does.

**Review by krshdev — COMMENTED** [review]
I looked up the documentation for this, and I had a question, did you have a look at these? 

image

**Comment by Pranav016** [comment]
> I looked up the documentation for this, and I had a question, did you have a look at these?
> 
> image

It only breaks if we are never updating the cache(which usually happens with our local cache) but in this case we are updating the cache whenever our lock file is changed and we store cache even for different node versions if we decide to test on multiple. I have cached my dependencies in the past, even with `npm ci` and they haven't given any issues..
Can you also provide a link to the reference docs, I couldn't find the dedicated note.

**Comment by krshdev** [comment]
> Can you also provide a link to the reference docs, I couldn't find the dedicated note.

https://github.com/actions/cache/blob/main/examples.md#node---npm

**Comment by nhcarrigan** [comment]
That's a good catch @krshdev! It sounds like this isn't an approach we want to take, then.

**Comment by nhcarrigan** [comment]
Going to close this per Eddie's comment above. Thanks for your interest in contributing!!!

**Pull request closed without merging**